### PR TITLE
SCA: Create the PaymentIntent just when the checkout form is submitted 

### LINKED ISSUE #787: SCA: Create the PaymentIntent just when the checkout form is submitted
See this comment: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/781/files#r257760521

Basically, we shouldn't create a new PaymentIntent every time a user opens the Checkout page, because all those intents are persisted and displayed in the merchant's Stripe dashboard.

Right after the user clicks the `Submit` button at checkout, we can create the Intent using an AJAX endpoint, and then injecting the Intent ID before continuing with the flow.

### PR DESCRIPTION
**Pull request by RadoslavGeorgiev**
Fixes #787.

#### Changes proposed in this Pull Request:

This PR introduces the `WC_Stripe_Intent_Generator`, which uses the new `get_intent_and_secret` method of `WC_Gateway_Stripe` in order to only generate payment intents when the checkout form is being submitted.

With this change, payment intents will not be generated on every visit of the checkout page. This way merchants will not see empty payment intents in their dashboards and the checkout page loads faster.

To prevent unintended direct access to the new AJAX endpoint, `WC_Stripe_Intent_Generator` will check if the cart has contents, and only create a payment intent when the cart is not empty. This is why a standard WC_AJAX action is used, instead of a REST API endpoint.

#### Testing instructions:

Try creating an order with a credit card. The process should start with a local AJAX request before Stripe's `handleCardPayment` is called.

### CODE REVIEW

File: includes/class-wc-gateway-stripe.php [hunk 1]
```diff
@@ -506,6 +487,7 @@ public function payment_scripts() {
 		$stripe_params['no_prepaid_card_msg']                     = __( 'Sorry, we\'re not accepting prepaid cards at this time. Your credit card has not been charged. Please try with alternative payment method.', 'woocommerce-gateway-stripe' );
 		$stripe_params['no_sepa_owner_msg']                       = __( 'Please enter your IBAN account name.', 'woocommerce-gateway-stripe' );
 		$stripe_params['no_sepa_iban_msg']                        = __( 'Please enter your IBAN account number.', 'woocommerce-gateway-stripe' );
+		$stripe_params['payment_intent_error']                    = __( 'We could not initiate the payment. Please try again!', 'woocommerce-gateway-stripe' );
```
Comment (DanReyLop) [inline includes/class-wc-gateway-stripe.php]: The other error messages are less enthusiastic but less formal, let's be consistent :)
```suggestion
		$stripe_params['payment_intent_error']                    = __( 'We couldn\'t initiate the payment. Please try again.', 'woocommerce-gateway-stripe' );
```

File: includes/class-wc-gateway-stripe.php [hunk 2]
```diff
@@ -1015,4 +997,35 @@ public function get_intent_object( $intent_id = '' ) {
...
+			'currency'             => get_woocommerce_currency(),
+			'allowed_source_types' => array(
+				'card',
+			),
+		);
+
+		$intent = WC_Stripe_API::request( $request, 'payment_intents' );
+
+		return array(
+			'id'            => $intent->id,
```
Comment (DanReyLop) [inline includes/class-wc-gateway-stripe.php]: I don't see the `id` property used anywhere. Shouldn't we keep a reference to the intent somewhere in the server?

File: assets/js/stripe.js [hunk 3]
```diff
@@ -494,10 +495,17 @@ jQuery( function( $ ) {
 			}
 
 			if ( 'card' === source_type ) {
-				stripe.handleCardPayment( client_secret, stripe_card, {
-					source_data: extra_details,
-				} )
-				.then( wc_stripe_form.intentResponse );
+				wc_stripe_form.getIntent()
+					.catch( function( error ) {
```
Comment (DanReyLop) [inline assets/js/stripe.js]: The `.catch()` should be after the `.then()` steps. As it is right now, if there's an error the `.catch()` will run and after the `.then()` will be called with no parameters so `intent.client_secret` will crash.

File: includes/class-wc-gateway-stripe.php (hunk 2 above)
Comment (RadoslavGeorgiev) [reply includes/class-wc-gateway-stripe.php]: More than that, we should use webhooks. This is a bigger task, so I created an issue (#)[https://github.com/woocommerce/woocommerce-gateway-stripe/issues/804], which deserves a separate PR.

**Pull request merged**